cartographer/io/serialization_format_migration.cc

```diff
@@ -47,46 +49,178 @@ bool ReadBuilderOptions(
+      submap_in.submap_3d().high_resolution_hybrid_grid();
+  *submap_3d.mutable_low_resolution_hybrid_grid() =
+      submap_in.submap_3d().low_resolution_hybrid_grid();
+
+  mapping::proto::Submap submap_out;
+  *submap_out.mutable_submap_3d() = submap_3d;
+  *submap_out.mutable_submap_id() = submap_in.submap_id();
+  return submap_out;
+}
+
 bool DeserializeNext(cartographer::io::ProtoStreamReaderInterface* const input,
                      ProtoMap* proto_map) {
   mapping::proto::LegacySerializedData legacy_data;
   if (!input->ReadProto(&legacy_data)) return false;
 
   if (legacy_data.has_submap()) {
+    LOG_FIRST_N(INFO, 1) << "Migrating submap data.";
+    if (legacy_data.submap().has_submap_2d()) {
+      CHECK(legacy_data.submap().submap_2d().grid().has_probability_grid_2d() ||
+            legacy_data.submap().submap_2d().grid().has_tsdf_2d());
+    }
     auto& output_vector = (*proto_map)[SerializedData::kSubmapFieldNumber];
     output_vector.emplace_back();
     *output_vector.back().mutable_submap() = legacy_data.submap();
   }
   if (legacy_data.has_node()) {
+    LOG_FIRST_N(INFO, 1) << "Migrating node data.";
+    auto& output_vector = (*proto_map)[SerializedData::kNodeFieldNumber];
+    output_vector.emplace_back();
+    *output_vector.back().mutable_node() = legacy_data.node();
+  }
+  if (legacy_data.has_trajectory_data()) {
+    LOG_FIRST_N(INFO, 1) << "Migrating trajectory data.";
+    auto& output_vector =
+        (*proto_map)[SerializedData::kTrajectoryDataFieldNumber];
+    output_vector.emplace_back();
+    *output_vector.back().mutable_trajectory_data() =
+        legacy_data.trajectory_data();
+  }
+  if (legacy_data.has_imu_data()) {
+    LOG_FIRST_N(INFO, 1) << "Migrating IMU data.";
+    auto& output_vector = (*proto_map)[SerializedData::kImuDataFieldNumber];
+    output_vector.emplace_back();
+    *output_vector.back().mutable_imu_data() = legacy_data.imu_data();
+  }
+  if (legacy_data.has_odometry_data()) {
+    LOG_FIRST_N(INFO, 1) << "Migrating odometry data.";
+    auto& output_vector = (*proto_map)[SerializedData::kOdometryData];
+    output_vector.emplace_back();
+    *output_vector.back().mutable_odometry_data() = legacy_data.odometry_data();
+  }
+  if (legacy_data.has_fixed_frame_pose_data()) {
+    LOG_FIRST_N(INFO, 1) << "Migrating fixed frame pose data.";
+    auto& output_vector =
+        (*proto_map)[SerializedData::kFixedFramePoseDataFieldNumber];
+    output_vector.emplace_back();
+    *output_vector.back().mutable_fixed_frame_pose_data() =
+        legacy_data.fixed_frame_pose_data();
+  }
+  if (legacy_data.has_landmark_data()) {
+    LOG_FIRST_N(INFO, 1) << "Migrating landmark data.";
+    auto& output_vector =
+        (*proto_map)[SerializedData::kLandmarkDataFieldNumber];
+    output_vector.emplace_back();
+    *output_vector.back().mutable_landmark_data() = legacy_data.landmark_data();
+  }
+  return true;
+}
+
+bool DeserializeNextWithLegacySubmap(
```

Comment:
instead of code dup'ing, please pass down a bool flag indicating that the submaps in the original stream are in the non-generalized grid format.

Reply:
I think this is not so trivial, because you would need to instantiate a "variant" type of `legacy_data` depending on this flag, i.e. `mapping::proto::LegacySerializedData legacy_data;` or `mapping::proto::LegacySerializedDataOldSubmap legacy_data;`. Templating is not possible because the methods of the protos don't have the same signatures (e.g. `legacy_data.submap().submap_2d().grid()` vs. `legacy_data.submap().submap_2d().probability_grid()`. This would only work for the part after   `if (legacy_data.has_node()) {`, which is indeed duplicated.

I thought about this, but didn't come up with a better or easy to read alternative. What do you think?

Reply:
True thing. Too bad ...